### PR TITLE
Adding initial client code for CGW (still work is in progrss for client)

### PR DESCRIPTION
**Pull request by jalam453**
Adding the client code for the initial review (NOTE: It is just a partial client code, work is still in progress for the client)

### CODE REVIEW

**Comment by jalam453** [comment]
@midnightercz Please review the CGW Client code

File: pubtools/_content_gateway/cgw_client.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+class CGWClient:
+    def __init__(self, endpoint, username, password):
+        self.endpoint = endpoint.rstrip("/")
+        self.username = username
+        self.password = password
+        self.prefix_url = "/content-gateway/rest/admin"
+        self.header = self.get_request_headers()
+
+    def get_request_headers(self):
```
Comment (midnightercz) [inline pubtools/_content_gateway/cgw_client.py:14]: move this to cgwauth client as something like make_auth()

File: pubtools/_content_gateway/cgw_client.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        if not self.username or not self.password:
+            raise CGWClientError("Error: username / password not found !!")
+
+        user_pass_string = "%s:%s" % (self.username, self.password)
+        user_pass = b64encode(user_pass_string.encode("utf-8")).decode("ascii")
+        headers = {"Authorization": "Basic %s" % user_pass}
+        return headers
+
+    def call_cgw_api(self, method, url, params=None, data=None):
```
Comment (midnightercz) [inline pubtools/_content_gateway/cgw_client.py:23]: make this method private

File: pubtools/_content_gateway/cgw_client.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+from base64 import b64encode
+
+
+class CGWClient:
+    def __init__(self, endpoint, username, password):
+        self.endpoint = endpoint.rstrip("/")
+        self.username = username
+        self.password = password
+        self.prefix_url = "/content-gateway/rest/admin"
+        self.header = self.get_request_headers()
```
Comment (midnightercz) [inline pubtools/_content_gateway/cgw_client.py:12]: self.headers

File: pubtools/_content_gateway/cgw_client.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+        endpoint_url = "%s%s" % (self.endpoint, url)
+
+        # This print is for debugging purpose only. Will remove it in the final push.
+        print(f"URL is {endpoint_url}")
+        response = None
+        output = True
+        try:
+            if method == "GET":
+                response = requests.get(url=endpoint_url, params=params, verify=False, headers=self.header)
```
Comment (midnightercz) [inline pubtools/_content_gateway/cgw_client.py:47]: make verify configurable

File: pubtools/_content_gateway/cgw_client.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        if not response.ok:
+            raise CGWClientError("content gateway API returned error: \nstatus_code: %s, reason: %s, error: %s"
+                                    % (response.status_code, response.reason, response.text))
+
+        if response.text:
+            output = response.json()
+        return output
+
+    def get_products(self, params=None):
+        url = "%s/products" % self.prefix_url
```
Comment (midnightercz) [inline pubtools/_content_gateway/cgw_client.py:69]: move this to call_cgw_api

File: pubtools/_content_gateway/cgw_client.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+class CGWClientError(Exception):
+    """
+    Custom exception to handle Content Gateway Client errors
+    """
+    def __init__(self, message):
+        self.message = message
+
+
+def unit_test():
```
Comment (midnightercz) [inline pubtools/_content_gateway/cgw_client.py:88]: that doesn't tell much about what is actually tested here

File: pubtools/_content_gateway/cgw_client.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    username = ""
+    password = ""
+
+    endpoint = "https://content-gateway--stage.apps.ext-waf.mpp.preprod.iad2.dc.paas.redhat.com"
```
Comment (midnightercz) [inline pubtools/_content_gateway/cgw_client.py:92]: please use fake url and make response with 
https://requests-mock.readthedocs.io/en/latest/

File: requirements.txt [hunk 8]
```diff
@@ -1,21 +1,7 @@
...
 cffi==1.15.0
-charset-normalizer==2.0.12
-cryptography==36.0.2
-idna==3.3
-jsonschema==4.4.0
-monotonic==1.6
-more-executors==2.10.1
--e git+ssh://git@github.com/jalam453/pubtools-content-gateway.git@3c07c7c9e28ae18a28cdc68aecbaeaeff31df7fc#egg=pubtools_content_gateway
-pushcollector==1.3.0
+-e git+ssh://git@github.com/jalam453/pubtools-content-gateway.git@689ee2a97a3a76d22e61ec2ef22d6fd05ff5bb4a#egg=pubtools_content_gateway
```
Comment (midnightercz) [inline requirements.txt:2]: ?

File: requirements.txt [hunk 9]
```diff
@@ -1,21 +1,7 @@
...
 pycparser==2.21
-PyNaCl==1.5.0
-pyrsistent==0.18.1
-python-qpid-proton==0.33.0
-python2==1.2
-PyYAML==6.0
-requests==2.27.1
 six==1.16.0
 urllib3==1.26.9
+virtualenv==20.14.1
```
Comment (midnightercz) [inline requirements.txt:6]: I doubt this one is needed

File: setup.py [hunk 10]
```diff
@@ -20,6 +20,8 @@ def get_requirements():
     with open("requirements.txt") as f:
         return f.read().splitlines()
 
+INSTALL_REQUIRES = ['urllib3', 'six', 'requests', 'mock', 'pytest']
+# INSTALL_REQUIRES = get_requirements()
```
Comment (midnightercz) [inline setup.py:24]: Please use this one instead

File: pubtools/_content_gateway/cgw_client.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+        # This print is for debugging purpose only. Will remove it in the final push.
+        print(f"URL is {endpoint_url}")
+        response = None
+        output = True
+        try:
+            if method == "GET":
+                response = requests.get(url=endpoint_url, params=params, verify=False, headers=self.header)
+            elif method == "POST":
+                response = requests.post(url=endpoint_url, data=data, verify=False, headers=self.header)
```
Comment (midnightercz) [inline pubtools/_content_gateway/cgw_client.py:49]: You have CGWSession object to be used here instead of pure requests

**Review by midnightercz — CHANGES_REQUESTED** [review]

**Comment by jalam453** [comment]
Sure! I'll make the changes and will send the patch again. Thanks.

File: pubtools/_content_gateway/cgw_client.py [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+        # Check if CGW hostname is present
+        if not self.hostname:
+            raise CGWClientError("No content gateway hostname found")
+
+        # Check if correct method id passed
+        if method not in ["GET", "POST", "PATCH", "PUT", "DELETE"]:
+            raise CGWClientError("Wrong request method passed")
+
+        response = None
+        output = True
```
Comment (midnightercz) [inline pubtools/_content_gateway/cgw_client.py:31]: I don't understand why you set it to True. Method should return one type of result. In this case python dict. In the case there is nothing to return (like for DELETE), return empty dict.

File: pubtools/_content_gateway/cgw_client.py [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+    """
+
+    def __init__(self, message):
+        self.message = message
+
+
+class CGWClient:
+    def __init__(self, hostname, cgw_auth=None, verify=True):
+        self.hostname = hostname.rstrip("/")
+        self.cgw_session = CGWSession(self.hostname)
```
Comment (midnightercz) [inline pubtools/_content_gateway/cgw_client.py:17]: pass verify here

**Review by midnightercz — CHANGES_REQUESTED** [review]

**Review by midnightercz — CHANGES_REQUESTED** [review]

File: pubtools/_content_gateway/cgw_client.py [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+    def get_products(self, params=None):
+        resp_data = self.call_cgw_api("GET", '/products', params=params)
+        return resp_data
+
+    def get_product(self, pid, params=None):
+        endpoint = "/products/%s" % pid
+        resp_data = self.call_cgw_api("GET", endpoint, params=params)
+        return resp_data
+
+    def create_product(self, data=None):
```
[the remark this answers is not in the document]
Comment (querti) [reply pubtools/_content_gateway/cgw_client.py:62]: Also it should probably be mentioned that the data should be a dict and not a JSON

File: pubtools/_content_gateway/cgw_client.py [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+        if not response.ok:
+            raise CGWClientError("content gateway API returned error: "
+                                 "\nstatus_code: %s, reason: %s, error: %s"
+                                 % (response.status_code, response.reason, response.text))
+
+        if response.text:
+            output = response.json()
+        return output
+
+    def get_products(self, params=None):
```
Comment (querti) [inline pubtools/_content_gateway/cgw_client.py:53]: The extra params will be added as a dictionary? Seems like an odd convention

File: pubtools/_content_gateway/cgw_client.py [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+
+        response = None
+        output = {}
+        try:
+            if method == "GET":
+                response = self.cgw_session.get(endpoint, params=params)
+            elif method == "POST":
+                response = self.cgw_session.post(endpoint, data=data)
+            elif method == "PUT":
+                response = self.cgw_session.put(endpoint, data=data)
```
Comment (querti) [inline pubtools/_content_gateway/cgw_client.py:37]: There seems to be some inconsistency on where you pass extra params

File: pubtools/_content_gateway/cgw_session.py [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+        """HTTP get request against CGW server API
+
+        Args:
+            endpoint (str)
+                API specific endpoint for the request
+        Returns:
+            requests.Response
+        """
+
+        return self.session.get(self._api_url(endpoint), verify=self.verify, **kwargs)
```
Comment (querti) [inline pubtools/_content_gateway/cgw_session.py:48]: I wolder what would happen if someone decided to specify verify in kwargs

File: requirements.txt [hunk 18]
```diff
@@ -1,21 +1,5 @@
...
--e git+ssh://git@github.com/jalam453/pubtools-content-gateway.git@3c07c7c9e28ae18a28cdc68aecbaeaeff31df7fc#egg=pubtools_content_gateway
-pushcollector==1.3.0
 pycparser==2.21
-PyNaCl==1.5.0
-pyrsistent==0.18.1
-python-qpid-proton==0.33.0
-python2==1.2
-PyYAML==6.0
-requests==2.27.1
 six==1.16.0
```
Comment (querti) [inline requirements.txt:3]: Is there a particular reason for pinning specific versions?

File: setup.py [hunk 19]
```diff
@@ -45,10 +47,10 @@ def get_requirements():
     url="https://github.com/jalam453/pubtools-content-gateway",
     classifiers=classifiers,
     packages=find_packages(exclude=["tests"]),
-    install_requires=get_requirements(),
+    install_requires=INSTALL_REQUIRES,
```
Comment (querti) [inline setup.py:50]: Pls utilize the requirements file instead of specifying the dependencies in setup.py

File: requirements.txt [hunk 20]
```diff
@@ -1,21 +1,5 @@
...
 urllib3==1.26.9
```
Comment (querti) [inline requirements.txt:4]: Pretty sure that urllib is a dependency of requests

File: pubtools/_content_gateway/cgw_client.py [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+class CGWClientError(Exception):
+    """
+    Custom exception to handle Content Gateway Client errors
+    """
+
+    def __init__(self, message):
+        self.message = message
+
+
+class CGWClient:
```
[the remark this answers is not in the document]
Comment (jalam453) [reply pubtools/_content_gateway/cgw_client.py:13]: I have that in the plan to add docstrings for all of the methods but I'll add them together later in the development phase.

File: pubtools/_content_gateway/cgw_client.py [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+        # Check if correct method id passed
+        if method not in ["GET", "POST", "PATCH", "PUT", "DELETE"]:
+            raise CGWClientError("Wrong request method passed")
+
+        response = None
+        output = {}
+        try:
+            if method == "GET":
+                response = self.cgw_session.get(endpoint, params=params)
+            elif method == "POST":
```
[the remark this answers is not in the document]
Comment (jalam453) [reply pubtools/_content_gateway/cgw_client.py:34]: We Don't have any requirements for the Patch here. You can check the API document https://source.redhat.com/groups/public/cdx/middleware_engineering_services_mwes_wiki/redhat_developersjboss_download_manager_administration_ui

File: pubtools/_content_gateway/cgw_client.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jalam453) [reply pubtools/_content_gateway/cgw_client.py:62]: Good catch and I have updated the same.

File: pubtools/_content_gateway/cgw_client.py [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+        return resp_data
+
+    def create_product(self, data=None):
+        endpoint = "/products/"
+        resp_data = self.call_cgw_api("PUT", endpoint, data=json.dumps(data))
+        return resp_data
+
+    def update_product(self, data=None):
+        endpoint = "/products"
+        resp_data = self.call_cgw_api("POST", endpoint, data=json.dumps(data))
```
[the remark this answers is not in the document]
Comment (jalam453) [reply pubtools/_content_gateway/cgw_client.py:69]: Nop! As per the given doc it must be POST and even if I change it to PUT to update the contents it is returning a 405 error. So I am following the API doc only https://source.redhat.com/groups/public/cdx/middleware_engineering_services_mwes_wiki/redhat_developersjboss_download_manager_administration_ui

File: pubtools/_content_gateway/cgw_client.py [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+        return resp_data
+
+    def get_product(self, pid, params=None):
+        endpoint = "/products/%s" % pid
+        resp_data = self.call_cgw_api("GET", endpoint, params=params)
+        return resp_data
+
+    def create_product(self, data=None):
+        endpoint = "/products/"
+        resp_data = self.call_cgw_api("PUT", endpoint, data=json.dumps(data))
```
[the remark this answers is not in the document]
Comment (jalam453) [reply pubtools/_content_gateway/cgw_client.py:64]: I completely agreed but the APIs have been designed in a way that it is only accepting PUT to create and POST to update which is completely inverted behavior.

File: pubtools/_content_gateway/cgw_client.py [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+
+        # Check if correct method id passed
+        if method not in ["GET", "POST", "PATCH", "PUT", "DELETE"]:
+            raise CGWClientError("Wrong request method passed")
+
+        response = None
+        output = {}
+        try:
+            if method == "GET":
+                response = self.cgw_session.get(endpoint, params=params)
```
[the remark this answers is not in the document]
Comment (jalam453) [reply pubtools/_content_gateway/cgw_client.py:33]: Get is not accepting and extra query params.

File: pubtools/_content_gateway/cgw_client.py (hunk 16 above)
Comment (jalam453) [reply pubtools/_content_gateway/cgw_client.py:37]: Didn't get you.

File: pubtools/_content_gateway/cgw_session.py (hunk 17 above)
Comment (jalam453) [reply pubtools/_content_gateway/cgw_session.py:48]: That we have to handle separately.

File: requirements.txt (hunk 18 above)
Comment (jalam453) [reply requirements.txt:3]: Nop! that's just got created by pip freeze > requirements.txt command. Will update it later. For now, we are not using this file.

File: setup.py (hunk 19 above)
Comment (jalam453) [reply setup.py:50]: I am getting some issues while using requirements from files so I have mentioned the dependencies here in the setup file. Will be using the dependencies from the requirements file only but later in the development phase.

**Review by querti — APPROVED** [review]

**Pull request merged**